widgets/edd-widget-most-commented.php

```diff
@@ -1,265 +1,257 @@
-                    <label for="<?php echo $this->get_field_id( 'show_price' ); ?>"><?php _e( 'Display price?', 'edd-widgets-pack' ); ?></label> 
-                </p>                
-                <p>
-                    <input id="<?php echo $this->get_field_id( 'thumbnail' ); ?>" name="<?php echo $this->get_field_name( 'thumbnail' ); ?>" type="checkbox" value="1" <?php checked( '1', $thumbnail ); ?>/>
-                    <label for="<?php echo $this->get_field_id( 'thumbnail' ); ?>"><?php _e( 'Display thumbnails?', 'edd-widgets-pack' ); ?></label> 
-                </p>
-                <p>
-                    <label for="<?php echo $this->get_field_id( 'thumbnail_size' ); ?>"><?php _e( 'Size of the thumbnails, e.g. <em>80</em> = 80x80px', 'edd-widgets-pack' ); ?></label> 
-                    <input class="widefat" id="<?php echo $this->get_field_id( 'thumbnail_size' ); ?>" name="<?php echo $this->get_field_name( 'thumbnail_size' ); ?>" type="text" value="<?php echo $thumbnail_size; ?>" />
-                </p>
-            <?php
-        }
-
-    }
+						// get the post thumbnail.
+						if ( 1 === $thumbnail && has_post_thumbnail( $download->ID ) ) {
+							$post_thumbnail = get_the_post_thumbnail( $download->ID, $thumbnail_size, array( 'title' => esc_attr( $title_attr ) ) ) . "\n";
+							$out .= "<li class=\"widget-download-with-thumbnail\">\n";
+							$out .= sprintf( $link, get_permalink( $download->ID ), esc_attr( $title_attr ), 'widget-download-thumb', $post_thumbnail );
+						} else {
+							$out .= "<li>\n";
+						}
+
+						// append the download's title.
+						$out .= sprintf( $link, get_permalink( $download->ID ), esc_attr( $title_attr ), 'widget-download-title', $title );
+
+						// get the price.
+						if ( 1 === $show_price ) {
+							if ( edd_has_variable_prices( $download->ID ) ) {
+								$price = edd_price_range( $download->ID );
+							} else {
+								$price = edd_currency_filter( edd_get_download_price( $download->ID ) );
+							}
+							$out .= sprintf( "<span class=\"widget-download-price\">%s</span>\n", $price );
+						}
+
+						// finish this element.
+						$out .= "</li>\n";
+					}
+					// finish the list.
+					$out .= "</ul>\n";
+				}
+
+				// set the widget's containers.
+				$cache = $args['before_widget'] . $out . $args['after_widget'];
+
+				// store the result on a temporal transient.
+				set_transient( 'edd_widgets_most_commented', $cache );
+
+			}
+
+			echo $cache;
+
+		}
+
+		/**
+		 * Update
+		 *
+		 * @return   array
+		 * @since    1.0
+		 */
+		public function update( $new_instance, $old_instance ) {
+			$instance = $old_instance;
+
+			// sanitize title.
+			$instance['title'] = strip_tags( $new_instance['title'] );
+
+			// sanitize limit.
+			$instance['limit'] = ( (bool) preg_match( '/^\-?[0-9]+$/', $instance['limit'] ) );
```

Comment:
This is throwing a notice in the ajax--also it's casting it to a boolean. I think it needs to be something like this (typed in browser so don't trust me...also needs to end up allowing `-1` as a valid result):
```suggestion
			$instance['limit'] = ! empty( $instance['limit'] ) ? (int) $instance['limit'] : 4;
```

Reply:
Your suggested code works for the ajax error.  I wonder why the min would be -1. I'd love to change it to 0 if you don't think there would be any larger implications, since -1 doesn't really make sense in this case.

Reply:
This field allows `-1` so that someone can tell the widget to display ALL the downloads (which seems counterintuitive based on the widget), but that is what was allowed. Interestingly, the empty check would fail if someone entered `0` in the field, which also seems counterintuitive, but do we need to allow `0` to be a value for it? If so, it should be an `isset` check, not an `empty` check.

Reply:
I think you're right with changing it to `isset`.  `0` should evaluate to true for both, but it looks like `-1` is only true with `isset`.
Source: https://www.php.net/manual/en/types.comparisons.php